src/Storages/Statistics/ConditionSelectivityEstimator.cpp

```diff
@@ -135,38 +171,201 @@ void ConditionSelectivityEstimator::ColumnSelectivityEstimator::addStatistics(St
+            {
+                out.function = RPNElement::FUNCTION_IN_RANGE;
+                out.column_not_ranges.emplace(column, Range(value));
+            }
+        },
+        {
+            "equals",
+            [] (RPNElement & out, const String & column, const Field & value)
+            {
+                out.function = RPNElement::FUNCTION_IN_RANGE;
+                out.column_ranges.emplace(column, Range(value));
+            }
+        },
+        {
+            "in",
+            [] (RPNElement & out, const String & column, const Field & value)
+            {
+                out.function = RPNElement::FUNCTION_IN_RANGE;
+                Ranges ranges;
+                for (const Field & field : value.safeGet<Tuple>())
+                {
+                    ranges.emplace_back(field);
+                }
+                out.column_ranges.emplace(column, PlainRanges(ranges, /*intersect*/ true, /*ordered*/ false));
+            }
+        },
+        {
+            "notIn",
+            [] (RPNElement & out, const String & column, const Field & value)
+            {
+                out.function = RPNElement::FUNCTION_IN_RANGE;
+                Ranges ranges;
+                for (const Field & field : value.safeGet<Tuple>())
+                {
+                    ranges.emplace_back(field);
+                }
+                out.column_not_ranges.emplace(column, PlainRanges(ranges, /*intersect*/ true, /*ordered*/ false));
+            }
+        },
+        {
+            "less",
+            [] (RPNElement & out, const String & column, const Field & value)
+            {
+                out.function = RPNElement::FUNCTION_IN_RANGE;
+                out.column_ranges.emplace(column, Range::createRightBounded(value, false));
+            }
+        },
+        {
+            "greater",
+            [] (RPNElement & out, const String & column, const Field & value)
+            {
+                out.function = RPNElement::FUNCTION_IN_RANGE;
+                out.column_ranges.emplace(column, Range::createLeftBounded(value, false));
+            }
+        },
+        {
+            "lessOrEquals",
+            [] (RPNElement & out, const String & column, const Field & value)
+            {
+                out.function = RPNElement::FUNCTION_IN_RANGE;
+                out.column_ranges.emplace(column, Range::createRightBounded(value, true));
+            }
+        },
+        {
+            "greaterOrEquals",
+            [] (RPNElement & out, const String & column, const Field & value)
+            {
+                out.function = RPNElement::FUNCTION_IN_RANGE;
+                out.column_ranges.emplace(column, Range::createLeftBounded(value, true));
+            }
+        }
+};
+
+/// merge CNF or DNF
+bool ConditionSelectivityEstimator::RPNElement::tryToMergeClauses(RPNElement & lhs, RPNElement & rhs)
 {
-    return rows - estimateLess(val, rows);
+    auto canMergeWith = [&](const RPNElement & e)
```

Comment:
```suggestion
    auto canMergeWith = [](const RPNElement & e)
```
Seems like we don't need capture